### PR TITLE
102 add sad parameter

### LINKED ISSUE #102: Add SAD parameter for SH12A

### PR DESCRIPTION
**Pull request by hendzeld**
Resolves  #102

### CODE REVIEW

File: converter/shieldhit/beam.py [hunk 1]
```diff
@@ -67,6 +69,7 @@ class BeamConfig:
     multiple_scattering: MultipleScatteringMode = MultipleScatteringMode.MOLIERE
 
     energy_cutoff_template = "TCUT0 {energy_low_cutoff} {energy_high_cutoff}  ! energy cutoffs [MeV]"
+    beam_sad_template = "BEAMSAD {beam_sad_x} {beam_sad_y}  ! BEAMSAD value [cm]"
```
Comment (grzanka) [inline converter/shieldhit/beam.py:72]:
```suggestion
    beam_sad_template = "SAD {beam_sad_x} {beam_sad_y}  ! SAD value [cm]"
```

File: converter/shieldhit/beam.py [hunk 2]
```diff
@@ -140,11 +144,18 @@ def __str__(self) -> str:
                 energy_high_cutoff=self.energy_high_cutoff
             )
 
+        # if beamsad was defined, add it to the template
+        beam_sad_line = "! no BEAMSAD value"
```
Comment (grzanka) [inline converter/shieldhit/beam.py:148]:
```suggestion
        beam_sad_line = "! no SAD value"
```

File: tests/test_shieldhit_config_mappings.py [hunk 3]
```diff
@@ -13,6 +16,7 @@
 BEAMPOS 0 0 0 ! Position of the beam
 BEAMDIR 0.0 0.0 ! Direction of the beam
 BEAMSIGMA  -0.1 0.1  ! Beam extension
+! no BEAMSAD value
```
Comment (grzanka) [inline tests/test_shieldhit_config_mappings.py:19]:
```suggestion
! no SAD value
```

File: tests/test_shieldhit_config_mappings.py [hunk 4]
```diff
@@ -96,3 +100,20 @@ def test_energy_cutoff() -> None:
 
     beam.energy_high_cutoff = None
     assert "TCUT0" not in str(beam)
+
+@pytest.mark.parametrize('value_x, value_y',[
+    (None, None),
+    (2, None),
+    (1, 2),
+])
```
Comment (grzanka) [inline tests/test_shieldhit_config_mappings.py:108]: Make tests with more meaningful parameters, like 200 and 250.

File: tests/test_shieldhit_config_mappings.py [hunk 5]
```diff
@@ -96,3 +100,20 @@ def test_energy_cutoff() -> None:
...
+def test_beamsad_parameter(value_x: Optional[float], value_y: Optional[float]) -> None:
+    """Test if the BEAMSAD values are correctly set."""
+    beam = BeamConfig()
+    beam.beam_sad_x =  value_x
+    beam.beam_sad_y = value_y
+    if value_x is None and value_y is None:    
+        assert "! no BEAMSAD value" in str(beam)
+    elif value_x is not None and  value_y is None:
+        assert f"BEAMSAD {value_x}   ! BEAMSAD value [cm]" in str(beam)
```
Comment (grzanka) [inline tests/test_shieldhit_config_mappings.py:117]:
```suggestion
        assert f"SAD {value_x}   ! SAD value [cm]" in str(beam)
```

File: tests/test_shieldhit_config_mappings.py [hunk 6]
```diff
@@ -96,3 +100,20 @@ def test_energy_cutoff() -> None:
...
+    elif value_x is not None and value_y is not None:
+        assert f"BEAMSAD {value_x} {value_y}  ! BEAMSAD value [cm]" in str(beam)
```
Comment (grzanka) [inline tests/test_shieldhit_config_mappings.py:119]:
```suggestion
        assert f"SAD {value_x} {value_y}  ! SAD value [cm]" in str(beam)
```

**Comment by grzanka** [comment]
For tests I propose following procedure:
1. Launch UI which includes your PR with SAD support.
2. Create new project (it may be based on examples) and set SAD to some values (i.e. SADx = 200, SADy=250)
3. Save project as JSON file
4. Add such JSON to test resources in converter repo
5. Add some tests based on parsing such JSON
6. Run manual tests with `python converter/main.py editor.json workspace`, inspect generated files (i.e. `beam.dat`)
7. Run SHIELD-HIT12A (by calling `shieldhit` binary) on generated files and check if simulation is not crashing

File: tests/test_shieldhit_config_mappings.py (hunk 4 above)
Comment (hendzeld) [reply tests/test_shieldhit_config_mappings.py]: Done.

**Comment by grzanka** [comment]
I apologize for reverting the change. We have an ongoing MR in shieldhit repository, which removes SAD and makes as default BEAMSAD.

image

Can you adjust accordingly the converter ?

**Comment by hendzeld** [comment]
> I apologize for reverting the change. We have an ongoing MR in shieldhit repository, which removes SAD and makes as default BEAMSAD.
> 
> 
> Can you adjust accordingly the converter ?

Yes, so if I understand correctly I should use `BEAMSAD` instead `SAD` in generated `beam.dat` file.

Do we want to rewrite it also in JSON files generated by editor in UI?
Or maybe we will leave it as `sad` there?

**Comment by grzanka** [comment]
> > I apologize for reverting the change. We have an ongoing MR in shieldhit repository, which removes SAD and makes as default BEAMSAD.
> > Can you adjust accordingly the converter ?
> 
> Yes, so if I understand correctly I should use `BEAMSAD` instead `SAD` in generated `beam.dat` file.
> 
> Do we want to rewrite it also in JSON files generated by editor in UI? Or maybe we will leave it as `sad` there?

I would keep JSON as it is and adjust only generated `beam.dat` file

**Review by grzanka — APPROVED** [review]

**Pull request merged**